File: smart_screens/smart_screens/page/shift_quantity_planned_report/shift_quantity_planned_report.js
frappe.pages['shift-quantity-planned-report'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Shift Quantity Planned Report',
        single_column: true
    });

    // Initialize the report
    new ShiftQuantityPlannedReport(page);
};

class ShiftQuantityPlannedReport {
    constructor(page) {
        this.page = page;
        this.wrapper = $(this.page.body);
        this.filters = {};
        this.data = [];
        
        this.init();
    }

    init() {
        this.setup_page();
        this.setup_filters();
        this.setup_events();
        this.load_filter_options();
        this.load_data();
    }

    setup_page() {
        // Load the HTML template
        this.wrapper.html(frappe.render_template('shift_quantity_planned_report'));
    }

    setup_filters() {
        // Set default date to today
        const today = frappe.datetime.get_today();
        $('#date-filter').val(today);
        this.filters.date = today;
    }

    setup_events() {
        const self = this;

        // Apply filters button
        $('#apply-filters').on('click', function() {
            self.apply_filters();
        });

        // Refresh button
        $('#refresh-data').on('click', function() {
            self.load_data();
        });

        // Export to Excel button
        $('#export-excel').on('click', function() {
            self.export_to_excel();
        });

        // Filter change events
        $('#date-filter, #shift-filter').on('change', function() {
            // Auto-apply filters on change
            setTimeout(() => self.apply_filters(), 300);
        });
    }

    async load_filter_options() {
        try {
            const response = await frappe.call({
                method: 'smart_screens.smart_screens.page.shift_quantity_planned_report.shift_quantity_planned_report.get_filter_options'
            });

            if (response.message) {
                this.populate_shift_filter(response.message.shift_types);
            }
        } catch (error) {
            console.error('Error loading filter options:', error);
        }
    }

    populate_shift_filter(shift_types) {
        const shiftSelect = $('#shift-filter');
        shiftSelect.empty().append('<option value="">All Shifts</option>');
        
        shift_types.forEach(shift => {
            shiftSelect.append(`<option value="${shift.name}">${shift.name}</option>`);
        });
    }

    apply_filters() {
        this.filters = {
            date: $('#date-filter').val(),
            shift_type: $('#shift-filter').val()
        };
        
        this.load_data();
    }

    async load_data() {
        this.show_loading();
        this.hide_error();

        try {
            const response = await frappe.call({
                method: 'smart_screens.smart_screens.page.shift_quantity_planned_report.shift_quantity_planned_report.get_shift_quantity_planned_data',
                args: {
                    filters: this.filters
                }
            });

            if (response.message) {
                if (response.message.error) {
                    this.show_error(response.message.error);
                } else {
                    this.data = response.message.data || [];
                    this.render_data();
                    this.update_summary();
                }
            }
        } catch (error) {
            console.error('Error loading data:', error);
            this.show_error('Failed to load data. Please try again.');
        } finally {
            this.hide_loading();
        }
    }

    render_data() {
        const tbody = $('#data-tbody');
        tbody.empty();

        if (this.data.length === 0) {
            this.show_no_data();
            return;
        }

        this.hide_no_data();

        this.data.forEach(row => {
            const tr = $(`
                <tr>
                    <td>${row.item || ''}</td>
                    <td>${row.press || ''}</td>
                    <td>${row.mould || ''}</td>
                    <td>${row.date ? frappe.datetime.str_to_user(row.date) : ''}</td>
                    <td>${row.shift_type || ''}</td>
                    <td class="text-right">${this.format_number(row.noof_cavities)}</td>
                    <td class="text-right">${this.format_number(row.target_qty)}</td>
                    <td class="text-right font-weight-bold text-primary">${this.format_number(row.expected_production_qty)}</td>
                    <td><span class="badge badge-info">${row.plan_types_str || 'N/A'}</span></td>
                </tr>
            `);
            tbody.append(tr);
        });
    }

    update_summary() {
        if (this.data.length === 0) {
            $('#summary-cards').hide();
            return;
        }

        const totalItems = new Set(this.data.map(row => row.item)).size;
        const totalPresses = new Set(this.data.map(row => row.press)).size;
        const totalExpectedQty = this.data.reduce((sum, row) => sum + (row.expected_production_qty || 0), 0);
        const totalRecords = this.data.length;

        $('#total-items').text(totalItems);
        $('#total-presses').text(totalPresses);
        $('#total-expected-qty').text(this.format_number(totalExpectedQty));
        $('#total-records').text(totalRecords);

        $('#summary-cards').show();
    }

    export_to_excel() {
        if (this.data.length === 0) {
            frappe.msgprint('No data to export');
            return;
        }

        // Prepare data for export
        const exportData = this.data.map(row => ({
            'Item': row.item || '',
            'Press': row.press || '',
            'Mould': row.mould || '',
            'Date': row.date ? frappe.datetime.str_to_user(row.date) : '',
            'Shift Type': row.shift_type || '',
            'No. of Cavities': row.noof_cavities || 0,
            'Target Qty (Lifts)': row.target_qty || 0,
            'Expected Production Qty': row.expected_production_qty || 0,
            'Plan Types': row.plan_types_str || ''
        }));

        // Create filename with current timestamp
        const timestamp = frappe.datetime.now_datetime().replace(/[: ]/g, '_');
        const filename = `Shift_Quantity_Planned_Report_${timestamp}`;

        // Export to Excel
        frappe.tools.downloadify(exportData, null, filename);
        frappe.show_alert({
            message: 'Report exported successfully',
            indicator: 'green'
        });
    }

    format_number(value) {
        if (value === null || value === undefined || value === '') {
            return '0';
        }
        return parseFloat(value).toLocaleString('en-IN', {
            minimumFractionDigits: 0,
            maximumFractionDigits: 2
        });
    }

    show_loading() {
        $('#loading-indicator').show();
        $('#data-table').parent().hide();
        $('#summary-cards').hide();
    }

    hide_loading() {
        $('#loading-indicator').hide();
        $('#data-table').parent().show();
    }

    show_error(message) {
        $('#error-message').text(message).show();
    }

    hide_error() {
        $('#error-message').hide();
    }

    show_no_data() {
        $('#no-data-message').show();
        $('#summary-cards').hide();
    }

    hide_no_data() {
        $('#no-data-message').hide();
    }
}

// Make the class globally available for debugging
window.ShiftQuantityPlannedReport = ShiftQuantityPlannedReport;

File: smart_screens/smart_screens/page/shift_quantity_planned_report/shift_quantity_planned_report.py
import frappe
from frappe import _
import json
from datetime import datetime

@frappe.whitelist()
def get_shift_quantity_planned_data(filters=None):
    """
    Get shift quantity planned report data
    Expected Production Qty = Target No. Of Lifts × No. Of Cavities
    """
    try:
        # Convert filters string to dict if provided
        if filters and isinstance(filters, str):
            filters = json.loads(filters)
        
        # Initialize filters
        date_filter = filters.get('date') if filters else None
        shift_filter = filters.get('shift_type') if filters else None
        
        # Build conditions
        conditions = []
        values = {}
        
        if date_filter:
            conditions.append("wp.date = %(date)s")
            values['date'] = date_filter
            
        if shift_filter:
            conditions.append("wp.shift_type = %(shift_type)s")
            values['shift_type'] = shift_filter
        
        where_clause = " AND ".join(conditions) if conditions else "1=1"
        
        # Query for regular work planning data
        regular_query = f"""
            SELECT 
                wpi.item,
                wpi.work_station as press,
                wpi.mould,
                wp.date,
                wp.shift_type,
                ms.noof_cavities,
                wpit.target_qty,
                CASE 
                    WHEN ms.noof_cavities IS NOT NULL AND wpit.target_qty IS NOT NULL 
                    THEN CAST(ms.noof_cavities AS FLOAT) * CAST(wpit.target_qty AS FLOAT)
                    ELSE 0
                END as expected_production_qty,
                'Regular' as plan_type
            FROM `tabWork Planning` wp
            INNER JOIN `tabWork Plan Item` wpi ON wp.name = wpi.parent
            LEFT JOIN `tabMould Specification` ms ON wpi.mould = ms.mould_ref AND ms.docstatus = 1
            LEFT JOIN `tabShift Type` st ON wp.shift_type = st.name
            LEFT JOIN `tabWork Plan Item Target` wpit ON wpi.item = wpit.item AND TIME(st.total_time) = wpit.shift_type
            WHERE wp.docstatus = 1 AND {where_clause}
        """
        
        # Query for add-on work planning data
        addon_query = f"""
            SELECT 
                awpi.item,
                awpi.work_station as press,
                awpi.mould,
                awp.date,
                awp.shift_type,
                ms.noof_cavities,
                wpit.target_qty,
                CASE 
                    WHEN ms.noof_cavities IS NOT NULL AND wpit.target_qty IS NOT NULL 
                    THEN CAST(ms.noof_cavities AS FLOAT) * CAST(wpit.target_qty AS FLOAT)
                    ELSE 0
                END as expected_production_qty,
                'Add-on' as plan_type
            FROM `tabAdd On Work Planning` awp
            INNER JOIN `tabAdd On Work Plan Item` awpi ON awp.name = awpi.parent
            LEFT JOIN `tabMould Specification` ms ON awpi.mould = ms.mould_ref AND ms.docstatus = 1
            LEFT JOIN `tabShift Type` st2 ON awp.shift_type = st2.name
            LEFT JOIN `tabWork Plan Item Target` wpit ON awpi.item = wpit.item AND TIME(st2.total_time) = wpit.shift_type
            WHERE awp.docstatus = 1 AND {where_clause.replace('wp.', 'awp.')}
        """
        
        # Execute queries
        regular_data = frappe.db.sql(regular_query, values, as_dict=True)
        addon_data = frappe.db.sql(addon_query, values, as_dict=True)
        
        # Combine data
        all_data = regular_data + addon_data
        
        # Group and aggregate by item and press
        aggregated_data = {}
        
        for row in all_data:
            key = f"{row.item}_{row.press}"
            
            if key not in aggregated_data:
                aggregated_data[key] = {
                    'item': row.item,
                    'press': row.press,
                    'mould': row.mould,
                    'date': row.date,
                    'shift_type': row.shift_type,
                    'noof_cavities': row.noof_cavities or 0,
                    'target_qty': row.target_qty or 0,
                    'expected_production_qty': 0,
                    'plan_types': []
                }
            
            # Aggregate expected production quantity
            aggregated_data[key]['expected_production_qty'] += row.expected_production_qty or 0
            
            # Track plan types
            if row.plan_type not in aggregated_data[key]['plan_types']:
                aggregated_data[key]['plan_types'].append(row.plan_type)
        
        # Convert to list and add plan types as string
        result = []
        for data in aggregated_data.values():
            data['plan_types_str'] = ', '.join(data['plan_types'])
            result.append(data)
        
        # Sort by item name
        result.sort(key=lambda x: x['item'])
        
        return {
            'data': result,
            'total_records': len(result)
        }
        
    except Exception as e:
        frappe.log_error(f"Error in get_shift_quantity_planned_data: {str(e)}")
        return {
            'data': [],
            'total_records': 0,
            'error': str(e)
        }

@frappe.whitelist()
def get_shift_types():
    """Get all available shift types"""
    try:
        shift_types = frappe.db.sql("""
            SELECT name, start_time, end_time 
            FROM `tabShift Type`
            ORDER BY name
        """, as_dict=True)
        
        return shift_types
    except Exception as e:
        frappe.log_error(f"Error in get_shift_types: {str(e)}")
        return []

@frappe.whitelist()
def get_filter_options():
    """Get filter options for the report"""
    try:
        # Get distinct dates from work planning
        dates_query = """
            SELECT DISTINCT date 
            FROM (
                SELECT date FROM `tabWork Planning` WHERE docstatus = 1
                UNION
                SELECT date FROM `tabAdd On Work Planning` WHERE docstatus = 1
            ) as combined_dates
            ORDER BY date DESC
            LIMIT 30
        """
        
        dates = frappe.db.sql(dates_query, as_dict=True)
        
        # Get shift types
        shift_types = get_shift_types()
        
        return {
            'dates': [d.date for d in dates],
            'shift_types': shift_types
        }
        
    except Exception as e:
        frappe.log_error(f"Error in get_filter_options: {str(e)}")
        return {
            'dates': [],
            'shift_types': []
        }
